stdlib/public/core/AutoDiff.swift

```diff
@@ -87,11 +87,16 @@ public extension Differentiable
   }
 }
 
+// FIXME: This is currently commented because the where clause leads to
+// associated type inference which conflicts with `Differentiable` derived
+// conformances.
+/*
```

Comment:
To expand on this: I had to comment this default implementation because the constraint `where TangentVector == CotangentVector` triggers associated type inference. Associated type inference fails (because `TangentVector == CotangentVector` is a recursive constraint and neither of them are concrete/specified) and prevents the `Differentiable` derived conformances code path from executing.

I discussed the problem with @rxwei and we agreed on commenting the extension as a workaround.
The consequence is that all "primitive" `Differentiable` types (like `FloatingPoint` and `Tensor<Scalar : FloatingPoint>`) need to define `tangentVector(from:)` even if `TangentVector == CotangentVector`.
It's not a big deal for `Differentiable` types that use derived conformances though, since `tangentVector(from:)` will be synthesized.

---

More info on the recursive constraint:
```console
# Output of `swift -Xllvm -debug-only="Associated type inference" struct_differentiable.swift`.
Inferring associated types from decl:
(func_decl "tangentVector(from:)" interface type='<Self where Self : Differentiable, Self.CotangentVector == Self.TangentVector, Self.TangentVector == Self.CotangentVector.CotangentVector.TangentVector, Self.CotangentVector.CotangentVector.TangentVector == Self.TangentVector.TangentVector.CotangentVector, Self.TangentVector.TangentVector.CotangentVector == Self.CotangentVector.TangentVector.TangentVector.CotangentVector, Self.CotangentVector.TangentVector.TangentVector.CotangentVector == Self.TangentVector.CotangentVector.CotangentVector.TangentVector, Self.TangentVector.CotangentVector.CotangentVector.TangentVector == Self.TangentVector.CotangentVector.TangentVector.CotangentVector> (Self) -> (Self.CotangentVector) -> Self.TangentVector' access=public
  (parameter "self" interface type='Self')
  (parameter_list
    (parameter "cotangent" apiName=from interface type='Self.CotangentVector')))
Considering whether TangentVector can infer to:
(dependent_member_type assoc_type=Swift.(file).Differentiable.TangentVector
  (base=struct_type decl=diff.(file).MyType@diff.swift:2:8))
++ we can same-type to:
(dependent_member_type assoc_type=Swift.(file).Differentiable.CotangentVector
  (base=struct_type decl=diff.(file).MyType@diff.swift:2:8))
++ seems legit
Considering whether CotangentVector can infer to:
(dependent_member_type assoc_type=Swift.(file).Differentiable.CotangentVector
  (base=struct_type decl=diff.(file).MyType@diff.swift:2:8))
++ we can same-type to:
(dependent_member_type assoc_type=Swift.(file).Differentiable.TangentVector
  (base=struct_type decl=diff.(file).MyType@diff.swift:2:8))
++ seems legit
Candidates for inference:

(Swift.(file).Differentiable.tangentVector(from:)
  (Swift.(file).Differentiable extension.tangentVector(from:)
    TangentVector := MyType.CotangentVector
    CotangentVector := MyType.TangentVector))

# It's problematic that associated type inference treats this default implementation
# as a candidate for inference. Inference fails, causing the derived conformances
# code path to get skipped.
```

[Small bit of relevant code:](https://github.com/apple/swift/blob/tensorflow/lib/Sema/TypeCheckProtocolInference.cpp#L1220)
```cpp
// From: `AssociatedTypeInference::findSolutionsRec`
// In: lib/Sema/TypeCheckProtocolInference.cpp.

// This code block here is the "associated type inference" code path.
// It triggers if associated type inference finds any type witness for `assocType`.

// If the extension `where TangentVector == CotangentVector` is not commented,
// `typeWitnesses.begin(assocType)` will not be empty.
// Note that if `typeWitnesses.begin(assocType)` is not empty, the
// `computeAbstractTypeWitness` code path below will not execute
// (because this code block either returns or continues an outer loop).
auto typeWitness = typeWitnesses.begin(assocType);
if (typeWitness != typeWitnesses.end()) {
  // The solution contains an error.
  if (typeWitness->first->hasError()) {
    recordMissing();
    return;
  }
  continue;
}

// `computeAbstractTypeWitness` involves the derived conformances code path.
// If `typeWitnesses.begin(assocType)` is not empty, this code will never run. :(
if (Type type = computeAbstractTypeWitness(assocType,
                                           /*allowDerived=*/true)) {
  ...
}
```

---

Fixing this is tricky. Associated type inference is fragile, I tried changing things (e.g. reordering code blocks) in `TypeCheckProtocolInference.cpp` but that caused stdlib compilation to fail.

I think a sane fix could be (in the code above) to attempt `computeAbstractTypeWitness` if all inferred type witnesses fail. I'll likely need to ask for help from associated type inference experts.